pkg/models/controllers/jobs.go

```diff
@@ -229,46 +234,80 @@ func deleteJob(namespace, job string) error {
 	return err
 }
 
+func (ctl *JobCtl) makeRevision(job *v1.Job) {
+	revisionIndex := -1
+	revisions, err := getRevisions(*job)
+
+	if err != nil {
+		glog.Error(err)
+		return
+	}
+
+	uid := job.UID
+	for index, revision := range revisions {
+		if revision.Uid == string(uid) {
+			currentRevision := getCurrentRevision(job)
+			if reflect.DeepEqual(currentRevision, revision) {
+				return
+			} else {
+				revisionIndex = index
+				break
+			}
+		}
+	}
+
+	if revisionIndex == -1 {
+		revisionIndex = len(revisions) + 1
+	}
+
+	revisions[revisionIndex] = getCurrentRevision(job)
+
+	revisionsByte, err := json.Marshal(revisions)
+	if err != nil {
+		glog.Error(err)
+	}
+
+	if job.Annotations == nil {
+		job.Annotations = make(map[string]string)
+	}
+
+	job.Annotations["revisions"] = string(revisionsByte)
+	ctl.K8sClient.BatchV1().Jobs(job.Namespace).Update(job)
+
+}
+
 func JobReRun(namespace, jobName string) (string, error) {
 	k8sClient = client.NewK8sClient()
 	job, err := k8sClient.BatchV1().Jobs(namespace).Get(jobName, metav1.GetOptions{})
 	if err != nil {
 		return "", err
 	}
+
 	newJob := *job
 	newJob.ResourceVersion = ""
 	newJob.Status = v1.JobStatus{}
 	newJob.ObjectMeta.UID = ""
+	newJob.Annotations["revisions"] = strings.Replace(job.Annotations["revisions"], Running, Unfinished, -1)
+
 	delete(newJob.Spec.Selector.MatchLabels, "controller-uid")
 	delete(newJob.Spec.Template.ObjectMeta.Labels, "controller-uid")
 
-	revisions, err := getRevisions(*job)
-
+	err = deleteJob(namespace, jobName)
 	if err != nil {
-		return "", err
-	}
-
-	index := len(revisions) + 1
-	value := getStatus(job)
-	revisions[index] = value
-
-	revisionsByte, err := json.Marshal(revisions)
-	if err != nil {
-		glog.Errorf("failed to rerun job %s, reason: %s", err, err)
+		glog.Errorf("failed to rerun job %s, reason: %s", jobName, err)
 		return "", fmt.Errorf("failed to rerun job %s", jobName)
 	}
 
-	newJob.Annotations["revisions"] = string(revisionsByte)
-
-	err = deleteJob(job.Namespace, job.Name)
-	if err != nil {
-		glog.Errorf("failed to rerun job %s, reason: %s", err, err)
-		return "", fmt.Errorf("failed to rerun job %s", jobName)
+	for i := 0; i < 3; i++ {
+		_, err = k8sClient.BatchV1().Jobs(namespace).Create(&newJob)
```

Comment:
It's better to add sleeping time before retrying and set retry times as a constant.

Reply:
The code has been modified according to your suggestion